pkg/crd/gen_integration_test.go

```diff
@@ -0,0 +1,224 @@
+				maxDescLen := 10
+				testCrdGeneration(crdGenParams{
+					rootPaths:    rootPaths,
+					expectedFile: expectedYamlPath(crdVersion, "maxdesclen"),
+					generator: crd.Generator{
+						CRDVersions: []string{crdVersion},
+						MaxDescLen:  &maxDescLen,
+					},
+				})
+			})
+		})
+
+		Context("using crd v1 beta1", func() {
+			crdVersion := "v1beta1"
+			Context("with default options", func() {
+				testCrdGeneration(crdGenParams{
+					rootPaths:    rootPaths,
+					expectedFile: expectedYamlPath(crdVersion, ""),
+					generator: crd.Generator{
+						CRDVersions: []string{crdVersion},
+					},
+				})
+			})
+			Context("with maxDescLen=10", func() {
+				maxDescLen := 10
+				testCrdGeneration(crdGenParams{
+					rootPaths:    rootPaths,
+					expectedFile: expectedYamlPath(crdVersion, "maxdesclen"),
+					generator: crd.Generator{
+						CRDVersions: []string{crdVersion},
+						MaxDescLen:  &maxDescLen,
+					},
+				})
+			})
+			Context("with trivialVersions=true", func() {
+				testCrdGeneration(crdGenParams{
+					rootPaths:    rootPaths,
+					expectedFile: expectedYamlPath(crdVersion, "trivial"),
+					generator: crd.Generator{
+						CRDVersions:     []string{crdVersion},
+						TrivialVersions: true,
+					},
+				})
+			})
+			Context("with preserveUnknownFields=false", func() {
+				preserveUnknownFields := false
+				testCrdGeneration(crdGenParams{
+					rootPaths:    rootPaths,
+					expectedFile: expectedYamlPath(crdVersion, "preserve_false"),
+					generator: crd.Generator{
+						CRDVersions:           []string{crdVersion},
+						PreserveUnknownFields: &preserveUnknownFields,
+					},
+				})
+			})
+		})
+
+		Context("when crdVersions=v1;v1beta1", func() {
+			testCrdGeneration(crdGenParams{
+				rootPaths:    rootPaths,
+				expectedFile: expectedYamlPath("v1v1beta1", ""),
+				generator: crd.Generator{
+					CRDVersions: []string{"v1", "v1beta1"},
+				},
+			})
+		})
+
+		Context("when crdVersions is empty (default to v1beta1)", func() {
+			testCrdGeneration(crdGenParams{
+				rootPaths:    rootPaths,
+				expectedFile: expectedYamlPath("v1beta1", ""),
+				generator: crd.Generator{
+					CRDVersions: []string{},
+				},
+			})
+		})
+	})
+})
+
+func testCrdGeneration(params crdGenParams) {
+	It("should generate the expected CRD", func() {
+		By("switching into testdata to appease go modules")
+		cwd, err := os.Getwd()
+		Expect(err).NotTo(HaveOccurred())
+		Expect(os.Chdir("./testdata")).To(Succeed()) // go modules are directory-sensitive
+		defer func() { Expect(os.Chdir(cwd)).To(Succeed()) }()
+
+		By("initializing the generator")
+		var gen genall.Generator = params.generator
+		rt, err := genall.Generators{&gen}.ForRoots(params.rootPaths)
+		Expect(err).NotTo(HaveOccurred())
+
+		outputDir, err := ioutil.TempDir("", "controller-tools-test")
+		Expect(err).NotTo(HaveOccurred())
+		defer os.RemoveAll(outputDir)
+		var buf bytes.Buffer
+		rt.OutputRules.Default = genall.OutputToBuffer{Buffer: &buf}
```

Comment:
If that makes sense we can probably adapt other integration tests to also not use temporary files.